### PR TITLE
VirtualMachineSnapshot API and Implementation

### PR DESCRIPTION
**Pull request by mhenriks**
**What this PR does / why we need it**:

API and controller/webhook for VirtualMachineSnapshots.

API is loosely base on VolumeSnapshot API described here:  https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-cis-volume-snapshot-beta/

Restore implementation forthcoming.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Depends on k8s 1.17 and a provisioner that implements beta snapshot API.

Feedback on the API definitions most important now.

### API

`VirtualMachineSnapshot`:  References the VirtualMachine to be snapshotted.  In the future, it may reference other stuff as well, like VirtualMachineInstances and maybe other stuff.  For now, this resource is for telling the system to create a VirtualMachineSnapshotContent resource based on the current state of a VirtualMachine.

`VirtualMachineSnapshotContent`:  Created by the snapshot controller.  Contains references to all resources required to restore a VirtualMachine.  Currently, that is the VirtualMachineSpec, PVC specs, and corresponding VolumeSnapshot names.  If the VolumeSnapshots specified do not exist, they will be created.

VolumeSnapshots and VolumeSnapshotContents are bound in a 1 to t relationship.  Similar to PVCs/PVs

`VirtualMachineRestore` (not in this PR): User specifies the VolumeSnapshot to use to restore a VirtualMachine.

### Components

**Snapshot Controller**

Basically, creates/populates VirtualMachineSnapshotContents resources from VirtualMachineSnapshots.  Like so:

1.  VirtualMachineSnapshot received
2.  "Lock" the source VM by adding a finalizer and setting `snapshotInProgress` in status
3.  Create VirtualMachineSnapshotContent containing VM spec, specs of any PVCs that are snapshot-able, and VolumeSnapshot names.
4.  Create VolumeSnapshots
5.  Wait for snapshots to complete
6.  Set "ready" status of VirtualMachineSnapshot
7.  Unlock source VM

**Webhook**

Validates VirtualMachineSnapshot specs, also check that VirtualMachines are not running at the time of creation. VirtualMachineSnapshot specs are immutable after initial creation.

**Test infra**

Bringing up k8s 1.17 provider will now start a hacked version of rook/ceph (on `make cluster-sync`) that supports the VolumeSnapshot beta API.  When rook/ceph properly supports the beta api, the manifests will be updated.

**Release note**:

```release-note
API and controller/webhook for VirtualMachineSnapshots
```

### CODE REVIEW

File: pkg/virt-controller/watch/snapshot.go [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+	}
+
+	if !ready {
+		creationTime = nil
+	}
+
+	contentCpy := content.DeepCopy()
+	contentCpy.Status.ReadyToUse = &ready
+	contentCpy.Status.CreationTime = creationTime
+	contentCpy.Status.Error = vmSnapshotError
```
Comment (davidvossel) [inline pkg/virt-controller/watch/snapshot.go:297]: here's a thought, since it's possible multiple errors could exist, it might be helpful to use conditions here which would let you express all the error conditions that might be occurring.

**Comment by mhenriks** [comment]
@davidvossel 

>I'm curious about why you decided to split the api into two objects, VMSnapshot and VMSnapshotContent. I know that's similar to what the snapshot API in k8s has done, but I don't have a great understanding of why and whether or not it makes sense for us to adopt the same pattern.

The main reason is for additional security.  The content resource contains the VM spec.  Which may contain sensitive info.  Imagine a service that snapshots every VM every hour.  It only needs permission to use snapshot resources.  It does not need access to the VMs.  It doesn't need access to the content.

File: pkg/virt-controller/watch/snapshot_base.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+		recorder:                    recorder,
+		resyncPeriod:                resyncPeriod,
+	}
+
+	vmSnapshotInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ctrl.handleVMSnapshot,
+			UpdateFunc: func(oldObj, newObj interface{}) { ctrl.handleVMSnapshot(newObj) },
+		},
+		ctrl.resyncPeriod,
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot_base.go:94]: In general, I think having a more frequent resync period can help deal with buggy controllers.  In this specific case, I want to update the status of a vmsnapshot when a volumesnapshot is deleted out of band.  There is a delete handler for volumesnapshots.  But can not always rely on getting it.  If the active controller crashes, for example.  So, the resync is there to make sure the status is reasonable accurate.

File: pkg/virt-controller/watch/snapshot.go [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+var currentTime = func() *metav1.Time {
+	t := metav1.Now()
+	return &t
+}
+
+func (ctrl *SnapshotController) updateVMSnapshot(vmSnapshot *vmsnapshotv1alpha1.VirtualMachineSnapshot) error {
+	log.Log.V(3).Infof("Updating VirtualMachineSnapshot %s/%s", vmSnapshot.Namespace, vmSnapshot.Name)
+
+	// Make sure status is initialized
+	if vmSnapshot.Status == nil {
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot.go:104]: If every member of a struct is a pointer, and the struct is not required, I think it makes sense for the struct to be a pointer as well.

File: pkg/virt-controller/watch/snapshot.go (hunk 1 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot.go:297]: Yeah I've been going back and forth on this

**Comment by mhenriks** [comment]
>interesting. That could make sense. What permissions would be necessary to restore a snapshot then?

There will be a VirtualMachineRestore resource that references a VirtualMachineSnapshot resource by name.  So whoever wants to do a restore will need permission to create VirtualMachineRestores.  I think that will be sufficient.  But could also have the VirtualMachineRestore webhook check that the user has permission to update the target VM.

File: pkg/virt-controller/watch/snapshot_base.go [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+		vmSnapshotContentInformer: vmSnapshotContentInformer,
+		storageClassInformer:      storageClassInformer,
+		pvcInformer:               pvcInformer,
+		vmInformer:                vmInformer,
+		crdInformer:               crdInformer,
+		recorder:                  recorder,
+		resyncPeriod:              resyncPeriod,
+	}
+
+	ctrl.dynamicInformerMap = map[string]*dynamicInformer{
```
Comment (davidvossel) [inline pkg/virt-controller/watch/snapshot_base.go:111]: what happens if someone attempts to create a vmsnapshot when the volume snapshot CRD is not registered yet?

I believe we had a similar issue with datavolumes. our validation webhook should be rejecting vmi/vm with datavolume if the datavolume crd isn't registered.

File: pkg/virt-controller/watch/snapshot.go [hunk 5]
```diff
@@ -0,0 +1,777 @@
...
+	return nil
+}
+
+func (ctrl *SnapshotController) updateVMSnapshotContent(content *vmssv1alpha1.VirtualMachineSnapshotContent) error {
+	log.Log.V(3).Infof("Updating VirtualMachineSnapshotContent %s/%s", content.Namespace, content.Name)
+
+	var volueSnapshotStatus []vmssv1alpha1.VolumeSnapshotStatus
+	var deletedSnapshots, skippedSnapshots []string
+
+	currentlyReady := content.Status != nil && content.Status.ReadyToUse != nil && *content.Status.ReadyToUse
```
Comment (davidvossel) [inline pkg/virt-controller/watch/snapshot.go:175]: this is a little different than how our other operators work. The status section here is actually influencing how the reconcile loop converges on the desired state, rather than the status existing simply to report the observed state. 

More commonly I'd expect this function to first detect the status of all the volume snapshots and create whatever ones don't exist. Then i'd expect the vmSnapshotContent's status to reflect the current observable state of all the volume snapshots.

so basically, I'd expect `currentlyReady` to reflect information we gather from volume snapshot informers as it relates to this VMSnapshotContent object, and not information already stored in the VMSnapshotContent object itself.

Does that make sense? The distinction seems kind of subtle but it's a fairly large change in expected behavior.

File: pkg/virt-controller/watch/snapshot.go [hunk 6]
```diff
@@ -0,0 +1,777 @@
...
+					"VolumeSnapshot %s no longer exists",
+					vsName,
+				)
+				deletedSnapshots = append(deletedSnapshots, vsName)
+				continue
+			}
+
+			if currentlyError {
+				log.Log.V(3).Infof("Not creating snapshot %s because in error state", vsName)
+				skippedSnapshots = append(skippedSnapshots, vsName)
```
Comment (davidvossel) [inline pkg/virt-controller/watch/snapshot.go:207]: When an error occurs, does that mean the snapshot will never recover? It appears that way since we no longer process volume snapshots once an error occurs.

File: pkg/virt-controller/watch/snapshot_base.go (hunk 4 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot_base.go:111]: > what happens if someone attempts to create a vmsnapshot when the volume snapshot CRD is not registered yet?

The snapshot will still get created.  We want to be able to snapshot VMs that don't reference PVCs.  The question is how to handle VMs with PVCs that cannot be snapshotted.  I think it is still okay to proceed with the snapshot.  Maybe the user only cares about backing up/restoring their VM config.  But maybe you/others don't agree?  If so, we can of course do something with a webhook.

File: pkg/virt-controller/watch/snapshot.go (hunk 5 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot.go:175]: I understand your point, and for the most part, what you describe is true.  The problem is that there is only a small window in time in which it is valid to create snapshots.  And that is before the content ever reaches "ready" or "error" state.  During this time, what you describe is true.  Once one of those states is reached, the source is unlocked and there are no more consistency guarantees so we should not take any more snapshots.

As a concrete example, let's say a snapshot is "ready" at t0.  Then an administrator deletes a VolumeSnapshot or a VolumeSnapshot goes into error state at t1.  We DO NOT want to create another VolumeSnapshot at t2.

File: pkg/virt-controller/watch/snapshot.go (hunk 6 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot.go:207]: For the most part, yes.  Would rather fail an operation than keep sources locked until success.

The one recoverable error is when a VolumSnapshot disappears and then reappears (without us recreating it).  This may happen if a storage administrator mistakenly deletes/restores a VolumeSnapshot.  Perhaps an unlikely scenario but something worth supporting I think.

**Comment by mhenriks** [comment]
attn @fabiand @rmohr @stu-gott @aglitke @awels your reviews would be appreciated

File: pkg/virt-controller/watch/snapshot.go (hunk 5 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/snapshot.go:186]: yep, makes perfect sense. prior states actually do matter for snapshots.

File: pkg/virt-controller/watch/snapshot.go (hunk 6 above)
Comment (davidvossel) [reply pkg/virt-controller/watch/snapshot.go:218]: yep, understood

**Comment by rmohr** [comment]
/approve

The direction looks good. 

@mhenriks could you file an issue regarding to making ceph pre-cached on kubevirtci as a follow-up?

@davidvossel I am ok with the PR. Could you do a final review and lgtm?

**Comment by mhenriks** [comment]
@rmohr issue file here:  https://github.com/kubevirt/kubevirtci/issues/381

Thanks for the review!

File: hack/builder/Dockerfile [hunk 7]
```diff
@@ -68,10 +68,6 @@ RUN \
     # Install defaulter-gen
     cd $GOPATH/src/k8s.io/code-generator/cmd/defaulter-gen && \
     git checkout v0.16.4 && \
-    go install && \
-     # Install client-gen
-    cd $GOPATH/src/k8s.io/code-generator/cmd/client-gen && \
```
Comment (davidvossel) [inline hack/builder/Dockerfile:73]: don't we still need this?!

**Comment by fabiand** [comment]
:rocket:

**Comment by haslersn** [comment]
> There will be a VirtualMachineRestore resource that references a VirtualMachineSnapshot resource by name. So whoever wants to do a restore will need permission to create VirtualMachineRestores. I think that will be sufficient. But could also have the VirtualMachineRestore webhook check that the user has permission to update the target VM.

Instead of having an existing target VM, the operator could also create the target VM. In that case, if you want to restore an existing VM, you first delete it and then post a `VirtualMachineRestore`. Any thoughts?

**Comment by mhenriks** [comment]
> > There will be a VirtualMachineRestore resource that references a VirtualMachineSnapshot resource by name. So whoever wants to do a restore will need permission to create VirtualMachineRestores. I think that will be sufficient. But could also have the VirtualMachineRestore webhook check that the user has permission to update the target VM.
> 
> Instead of having an existing target VM, the operator could also create the target VM. In that case, if you want to restore an existing VM, you first delete it and then post a `VirtualMachineRestore`. Any thoughts?

@haslersn yeah I think we could handle that case pretty easily.  We are also planning "clone vm" functionality that will make a new VM from a snapshot.  Which is slightly more involved but similar to what you propose.

**Comment by haslersn** [comment]
Why is a clone more involved than a conversion of an existing VM? Naively I would think it's the other way around: If you want to convert an existing VM, then not only do you have to handle the special case that the VM might be running, but also you have to decide what happens with PVCs that are attached to the VM.

**Comment by haslersn** [comment]
Why is a clone more involved than a conversion of an existing VM? Naively I would think it's the other way around: If you want to convert an existing VM, then not only do you have to handle the special case that the VM might be running, but also you have to decide what happens with PVCs that are attached to the VM.

**Comment by haslersn** [comment]
Unrelated question: Does a VM snapshot (currently or in the future) also capture the state of the VM such that execution can be resumed?

**Comment by rmohr** [comment]
> Unrelated question: Does a VM snapshot (currently or in the future) also capture the state of the VMI such that execution can be resumed?

Right now it is for offline snapshots only. Live snapshots is definitely something which we want in the future.
